Tasks/VsTest/VSTest.ts

```diff
@@ -0,0 +1,501 @@
+        defer.resolve(tempFile);
+    }); 
+    return defer.promise;
+}
+
+function setRunInParallellIfApplicable(vsVersion: number) {
+    if(runInParallel){
+        if(!isNaN(vsVersion) && vsVersion >= 14){  
+            var vs14Common = tl.getVariable("VS140COMNTools");
+            if ((vs14Common && pathExistsAsFile(path.join(vs14Common, "..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\TE.TestModes.dll"))) || vsVersion > 14) {
+                setRegistryKeyForParallelExecution(vsVersion);
+                return;
+            }
+        }
+    }
+}
+
+function resetRunInParallel(){
+    tl.warning("Install Visual Studio 2015 Update 1 or higher on your build agent machine to run the tests in parallel.");
+    runInParallel = false;
+}
+
+function locateVSVersion(): Q.Promise<number> {
+    var defer = Q.defer<number>();
+    var vsVersion = parseFloat(vsTestVersion);
+    if(!isNaN(vsVersion)){
+        defer.resolve(vsVersion);
+        return defer.promise;
+    }
+    
+    var regPath = "HKLM\\SOFTWARE\\Microsoft\\VisualStudio";
+    regedit.list(regPath).on('data', function(entry) {
+        if (entry && entry.data && entry.data.keys) {
+            var subkeys = entry.data.keys;
+            var versions = getFloatsFromStringArray(subkeys);
+            if (versions && versions.length > 0) {
+                versions.sort();
+                defer.resolve(parseFloat(versions[versions.length - 1]));
+                return defer.promise;
+            }
+        }
+        defer.resolve(null);        
+    });
+    return defer.promise;
+}
+
+function getFloatsFromStringArray(inputArray: string[]): string[] {
+    var outputArray: string[] = []
+    var count;
+    if (inputArray) {
+        for (count = 0; count < inputArray.length; count++) {
+            var floatValue = parseFloat(inputArray[count]);
+            if (!isNaN(floatValue)) {
+                outputArray.push(inputArray[count]);
+            }
+        }
+    }
+    return outputArray;
+}   
+
+function setRegistryKeyForParallelExecution(vsVersion: number){
+    var regKey = "HKCU\\SOFTWARE\\Microsoft\\VisualStudio\\" + vsVersion.toFixed(1) + "_Config\\FeatureFlags\\TestingTools\\UnitTesting\\Taef";
+    regedit.createKey(regKey, function(err){
+        if(!err){
+            var values = {
+                [regKey]: {
+                    'Value': {
+                         value: '1',
+                        type: 'REG_DWORD'
+                    }
+                }
+            };
+            regedit.putValue(values, function (err) { 
+                if(err){
+                    tl.warning("Error occured while setting registry key, Error: " + err);
+                }   
+            }); 
+        }else{
+            tl.warning("Error occured while creating registry key, Error: " + err);
+        }
+    });   
+}
+
+function readFileContents(filePath: string, encoding: string): Q.Promise<string> {
+    var defer = Q.defer<string>();
```

Comment:
I wonder if we should add to task lib.  That would allow for easy mocking via the json response file ...  Same for writeFile.  Seems like common things to do ...